### PR TITLE
raft_serverpb: trace peers' availability info

### PR DESCRIPTION
**Pull request by ethercflow**
Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

### CODE REVIEW

File: proto/raft_serverpb.proto [hunk 1]
```diff
@@ -117,10 +117,18 @@ enum ExtraMessageType {
...
+    bool miss_data = 1;
+    uint64 conf_ver = 2;
 }
 
 message ExtraMessage {
     ExtraMessageType type = 1;
     uint64 premerge_commit = 2;
-    repeated metapb.Peer check_peers = 3;
+    ExtraState extra_state = 3;
+    repeated metapb.Peer check_peers = 4;
```
Comment (disksing) [inline proto/raft_serverpb.proto:133]: no need to consider backward capatibility?

Comment (ethercflow) [reply proto/raft_serverpb.proto:133]: Should consider, thanks!

File: proto/raft_serverpb.proto [hunk 2]
```diff
@@ -117,10 +117,18 @@ enum ExtraMessageType {
     MsgHibernateRequest = 4;
     MsgHibernateResponse = 5;
     MsgRejectRaftLogCausedByMemoryUsage = 6;
+    MsgCollectPeerApplyIndexProgressRequest = 7;
```
Comment (BusyJay) [inline proto/raft_serverpb.proto:120]: Wrong name.

File: proto/raft_serverpb.proto [hunk 3]
```diff
@@ -117,10 +117,18 @@ enum ExtraMessageType {
...
+    MsgCollectPeerApplyIndexProgressResponse = 8;
+}
+
+message ExtraState {
```
Comment (BusyJay) [inline proto/raft_serverpb.proto:124]: Bad name.

File: proto/raft_serverpb.proto [hunk 4]
```diff
@@ -117,10 +117,17 @@ enum ExtraMessageType {
     MsgHibernateRequest = 4;
     MsgHibernateResponse = 5;
     MsgRejectRaftLogCausedByMemoryUsage = 6;
+    MsgCollectPeerApplyIndexProgressRequest = 7;
+    MsgCollectPeerApplyIndexProgressResponse = 8;
```
Comment (Connor1996) [inline proto/raft_serverpb.proto:121]: why need a response

File: proto/raft_serverpb.proto (hunk 3 above)
Comment (ethercflow) [reply proto/raft_serverpb.proto:124]: How about PeerExtraState?

File: proto/raft_serverpb.proto (hunk 4 above)
Comment (ethercflow) [reply proto/raft_serverpb.proto:121]: Oh, What I thought was that when the leader sends `MsgAppend` or `MsgHeartbeat`, it sends a `MsgCollectPeerApplyIndexProgressRequest`, and then other peers receive this information and return `MsgCollectPeerApplyIndexProgressResponse`.

But it should be that when non-leader peers send `MsgAppendResp` or `MsgHeartBeatResp`, they carry the `extra state` information, and when the leader receives `MsgAppendResp` or `MsgHeartBeatResp`, it processes the received information. So no need for an `extra resp type` right?

File: proto/raft_serverpb.proto (hunk 2 above)
Comment (ethercflow) [reply proto/raft_serverpb.proto]: Fixed, thanks!

File: proto/raft_serverpb.proto (hunk 4 above)
Comment (ethercflow) [reply proto/raft_serverpb.proto]: Removed, thanks!

**Review by Connor1996 — APPROVED** [review]
LGTM

File: proto/raft_serverpb.proto [hunk 5]
```diff
@@ -117,10 +117,13 @@ enum ExtraMessageType {
     MsgHibernateRequest = 4;
     MsgHibernateResponse = 5;
     MsgRejectRaftLogCausedByMemoryUsage = 6;
+    MsgTracePeerAvailabilityInfo = 7;
```
Comment (BusyJay) [inline proto/raft_serverpb.proto:120]: Can it be shorter?

Comment (ethercflow) [reply proto/raft_serverpb.proto:120]: I don't think it can, I looked at synonyms and there is no short :(

Comment (BusyJay) [reply proto/raft_serverpb.proto:120]: Does it need a request type and a response type?

Comment (ethercflow) [reply proto/raft_serverpb.proto:120]: No, this message will be sent by non-leader peers when replying `MsgAppendResp` or `MsgHeartBeatResp`, leader use this type to collect infos.

Comment (BusyJay) [reply proto/raft_serverpb.proto:120]: So it will be attached to every messages? I don't think it is necessary.

Comment (ethercflow) [reply proto/raft_serverpb.proto:120]: Yes, as every `MsgAppendResp` as applied index changes, or `MsgHeartBeatResp` if there's long time no write.

Comment (BusyJay) [reply proto/raft_serverpb.proto:128]: I don't object to report the response periodically, but I don't think it's necessary to be attached to heartbeat or append response. It should be queried on demand.

Comment (Connor1996) [reply proto/raft_serverpb.proto:128]: If it's queried on demand, it's still triggered on leader periodically. Seems unnecessary

Comment (BusyJay) [reply proto/raft_serverpb.proto:128]: Once it's recovered from not ready to ready and vice versa, then no need to query anymore until there the role is changed again.

File: proto/raft_serverpb.proto [hunk 6]
```diff
@@ -125,10 +125,13 @@ enum ExtraMessageType {
...
 }
 
 message ExtraMessage {
     ExtraMessageType type = 1;
     uint64 premerge_commit = 2;
     repeated metapb.Peer check_peers = 3;
+    bool miss_data = 4;
+    uint64 safe_ts = 5;
```
Comment (BusyJay) [inline proto/raft_serverpb.proto:136]: If it's not used, it should be removed.

Comment (ethercflow) [reply proto/raft_serverpb.proto]: Fixed, thanks!

**Comment by ethercflow** [comment]
@BusyJay Msg has been split into request and response, also added a new PeerState `PeerState::Unavailable` as discussed earlier in the group.

File: proto/raft_serverpb.proto [hunk 7]
```diff
@@ -106,6 +106,7 @@ enum PeerState {
     Applying = 1;
     Tombstone = 2;
     Merging = 3;
+    Unavailable = 4;
```
Comment (Connor1996) [inline proto/raft_serverpb.proto:109]: add a comment for this

**Review by Connor1996 — COMMENTED** [review]
rest LGTM

File: proto/raft_serverpb.proto [hunk 8]
```diff
@@ -139,10 +140,13 @@ enum ExtraMessageType {
     MsgHibernateRequest = 4;
     MsgHibernateResponse = 5;
     MsgRejectRaftLogCausedByMemoryUsage = 6;
+    MsgTracePeerAvailabilityInfoRequest = 7;
```
Comment (BusyJay) [inline proto/raft_serverpb.proto:143]:
```suggestion
    MsgAvailabilityRequest = 7;
```

Comment (ethercflow) [reply proto/raft_serverpb.proto]: Renamed, thanks!

**Review by BusyJay — APPROVED** [review]

**Pull request merged**